src/Kernel/Process.class.st

```diff
@@ -358,6 +358,14 @@ Process >> isActiveProcess [
 	^ self == Processor activeProcess
 ]
 
+{ #category : #testing }
+Process >> isBlocked [
+	"Answer true if blocked on a semaphore."
+	self isActiveProcess ifTrue: [ ^false ].
+	self isTerminated ifTrue: [ ^false ].
+	^myList class == Semaphore
```

Comment:
Funny I did not think that we are speaking about Semaphore case specifically. There can be any other blocking scenarious. And generally it is not only about blocking. Just think that we are running some long computation at the end (which may periodically sleep or not):
```Smalltalk
[1/0] fork.
[ 10 factorial ] repeat
```

Reply:
of course... baby steps :D I've just covered the most pressing case for the beginning ;)

Reply:
Hi Denis,

I've been thinking about your comments and has done some testing in Cuis to make sure I'm not presenting an idea that is only applicable in Pharo; it works :)

> I did not think that we are speaking about Semaphore case specifically. 

I'd like to clarify what I mean by "blocked": I'd like to consider it one of the classic 5 process states: active, runnable, suspended, terminated and waiting (= blocked); I chose #isBlocked to be compatible with Squeak/Cuis terminology. Because Pharo and Cuis implement Mutex using a Semaphore I think the only case when a process is waiting is when it is waiting on a semaphore. Please correct me if I missed some "waiting" scenario.

> There can be any other blocking scenarios. 

What is it you mean by blocking; what other blocking scenarios did you have in mind?

> And generally it is not only about blocking. Just think that we are running some long computation at the end (which may periodically sleep or not):
```
[1/0] fork.
[ 10 factorial ] repeat
```
Yes, this is a valid point; it'd be great to address this one as well; however, it seems to me it's a slightly different sort of a problem and I'd suggest to deal with it separately from the simpler "waiting" scenario. The point is in the example above the UI is not "idly" waiting for something but it's actively "busy" doing something. I expect a slightly different (or more general) approach might be needed, and hence I'd like to leave it aside for the moment.

In your example, I'd expect the UI just go on working; why should it stop and look around; any other process than UI wouldn't... or maybe the UI should spawn a new UI process in such a case, I don't have an opinion yet :) But what about in the following example, should the UI deal with the 1/0 error?
```
[1/0] fork.
[ 10000 factorial. Processor yield ] repeat
```
I was still inclined to say no but look at this example:
```
[1/0] fork.
[ 10000 factorial. 5 seconds asDelay wait ] repeat
```
Why keep the UI idle for 5 seconds and not let it deal with the 1/0 error in the meantime? (this PR addresses this scenario).

Or even this example:
```
[[1/0] fork.
[ 10000 factorial. Processor yield ] repeat] fork
```
Of course you'd expect the UI take care of the 1/0 error; and I believe the UI should ideally behave as any other process running the same example because the computation’s result shouldn’t depend on whether it runs in the UI or not because it's an implementation detail the user shouldn't need to know or care about.

So to summarize: I suggest to implement the "easy" part, i.e. spawn a new UI if the old UI is waiting on a semaphore, and deal with the "busy" UI scenario later. 

And of course, there are these other issues with "sub-cycles":
```
[1/0] fork.
5 seconds asDelay wait.
self confirm: 'test?'
```
and who knows what else... Why not find a clean solution for just the "waiting" UIs first :)

Reply:
> I was still inclined to say no but look at this example:
> 
> ```
> [1/0] fork.
> [ 10000 factorial. 5 seconds asDelay wait ] repeat
> ```
> 
> Why keep the UI idle for 5 seconds and not let it deal with the 1/0 error in the meantime? (this PR addresses this scenario).

Well, increase the priority of the forked process and PR will not help:
 ```Smalltalk
[1/0] forkAt: Processor activePriority + 1.
[ 10000 factorial. 5 seconds asDelay wait ] repeat
```

To be honest I more inclined now that this fix is not a good idea at all. My examples just demonstrate inconsistencies which will beat the user in a real life scenarios. And in the real life I have no idea what happens inside the code I am playing in the playground:
```Smalltalk
SomeNewCoolLib new execute
```
If the code does some multithreaded computations/networking it will wait at random time for the remote data and at random time it will do the computation when data received. Now if there is a bug then in some cases I can luckily get the debugger (when the background error is thrown exactly during the waiting period in the UI) and at the another time it will be just a locked UI (when the background error is thrown during a computation time in the UI). 
It means that the same code started in the workspace may occasionally show a different behaviour. I would prefer to always have a same one. Otherwise it is a breaking brain moment. 

So I would not distinguish the cases when UI is locked on semaphore or locked on computation. It does not matter IMHO - UI is just busy.

Another thought. All these problems we discussing here is about the user experience with workspace (playground). Maybe it should just never run code in the UI Process. Then the problem will just gone.

Reply:
That's a lot of brain food, thank you!

> Another thought. All these problems we discussing here is about the user experience with workspace (playground). Maybe it should just never run code in the UI Process. Then the problem will just gone.

Yes, it occurred to me when I learned the Playground examples are run "in the UI" but it is still beyond my skill level to try to analyze and implement the alternative; there must have been a reason Squeak and its forks (and VW too I guess) chose this strategy; it'd help to know why :)

Reply:
> I'd like to clarify what I mean by "blocked": I'd like to consider it one of the classic 5 process states: active, runnable, suspended, terminated and waiting (= blocked); I chose #isBlocked to be compatible with Squeak/Cuis terminology. Because Pharo and Cuis implement Mutex using a Semaphore I think the only case when a process is waiting is when it is waiting on a semaphore. Please correct me if I missed some "waiting" scenario.

I guess you are right about the current state of the image. There is implementation of another synchronization primitive but it is not integrated. See PrimitiveLock from https://github.com/pharo-ide/ReadWriteLock/tree/master/ReadWriteLock . 
Anyway it's a good idea to not hardcode the classes and rely on messages.

Reply:
> There is implementation of another synchronization primitive but it is not integrated. See PrimitiveLock from https://github.com/pharo-ide/ReadWriteLock/tree/master/ReadWriteLock .

Thanks for pointing this out; in that case it'd be better to implement #isBlocked using
```
isBlocked
	"Answer true if blocked on a condition variable.
	A process is blocked if it is waiting on some list
	(e.g. a Semaphore), other than the runnable process lists."

	myList ifNotNil: [ :list | ^list class ~~ ProcessList ].
	^false
```

or like in Squeak (by Eliot):
```
isBlocked
	"Answer true if blocked on a condition variable.
	 A process is blocked if it is waiting on some list
	 (e.g. a Semaphore), other than the runnable process lists."

	| myPriority |
	"Grab my priority now. Even though evaluation is strictly right-to-left,
	accessing Processor could involve a send."
	myPriority := priority.
	^myList
		ifNil: [false]
		ifNotNil: [:list | list ~~ (Processor waitingProcessesAt: myPriority)]
```

to be somewhat future-proof :)